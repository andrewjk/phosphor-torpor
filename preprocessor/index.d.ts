export type Preprocessor = (options: {
  content: string
  attributes: Record<string, string | boolean>
  markup: string
  filename?: string
}) => Processed | void | Promise<Processed | undefined>

export interface Processed {
  code: string
  map?: string | object
}

export interface PreprocessorGroup {
  name: string
  script?: Preprocessor
}

export function phosphorSvelteOptimize(): PreprocessorGroup
