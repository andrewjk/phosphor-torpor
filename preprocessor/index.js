import MagicString from "magic-string"

/**
 * @returns {import(".").PreprocessorGroup}
 */
export function phosphorSvelteOptimize() {
  return {
    name: "phosphor-svelte-optimize",
    script({ attributes, filename, content }) {
      if (!filename || /node_modules/.test(filename)) return

      const re = new RegExp(
        /import\s*{(?<imports>[^}]*)}\s*from\s*["']phosphor-svelte["']([\s\n]*;)?/g
      )

      let output = new MagicString(content, { filename })
      for (let match of content.matchAll(re)) {
        const modules = match.groups["imports"]
          .trim()
          .replace(/,$/, "")
          .split(",")
          .map((s) => s.trim())
        const icons = modules.filter((s) => !s.startsWith("type"))
        const types = modules.filter((s) => s.startsWith("type"))

        const newImports = icons.map((icon) => {
          let matches
          if ((matches = icon.match(/(\w+)\s+as\s+(\w+)/))) {
            const module = matches[1]
            const alias = matches[2]

            return `import ${alias} from "phosphor-svelte/lib/${module}";`
          }

          return `import ${icon} from "phosphor-svelte/lib/${icon}";`
        })

        if (types.length > 0) {
          const typeImports = types.map((t) => t.slice(5, t.length)).join(", ")
          newImports.push(
            `import type { ${typeImports} } from "phosphor-svelte";`
          )
        }

        output.update(
          match.index,
          match.index + match[0].length,
            newImports.join("\n")
        )
      }

      return {
        attributes,
        code: output.toString(),
        map: output.generateMap(),
      }
    },
  }
}
