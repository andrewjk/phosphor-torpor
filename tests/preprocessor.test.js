import { describe, expect, it } from "vitest"
import { phosphorSvelteOptimize } from "../preprocessor"
import { preprocess } from "svelte/compiler"

describe("preprocessor", () => {
  describe("phosphorSvelteOptimize", () => {
    it("should have a name", () => {
      const processor = phosphorSvelteOptimize()

      expect(processor).toHaveProperty("name", "phosphor-svelte-optimize")
    })

    it("should preprocess icon imports", async () => {
      const processor = phosphorSvelteOptimize()

      const content = `<script>
  import Airplane from "phosphor-svelte/lib/Airplane";
  import { Cube, Rectangle, Horse } from "phosphor-svelte";
</script>\n`

      const result = await preprocess(content, processor, {
        filename: "App.svelte",
      })
      expect(result).toMatchSnapshot()
    })

    it("should preprocess multiple imports", async () => {
      const processor = phosphorSvelteOptimize()

      const content = `<script>
  import { Airplane, Gear } from "phosphor-svelte";
  import { Cube, Rectangle, Horse } from "phosphor-svelte";
  import { X } from "phosphor-svelte";
</script>\n`

      const result = await preprocess(content, processor, {
        filename: "App.svelte",
      })
      expect(result).toMatchSnapshot()
    })

    it("should separate type imports", async () => {
      const processor = phosphorSvelteOptimize()

      const content = `<script lang="ts">
  import { Airplane, Gear, type IconContextProps } from "phosphor-svelte";
</script>\n`

      const result = await preprocess(content, processor, {
        filename: "App.svelte",
      })
      expect(result).toMatchSnapshot()
    })

    it("should preprocess aliased imports", () => {
      const processor = phosphorSvelteOptimize()

      const content = `<script>
  import { Airplane, Gear as IconGear } from "phosphor-svelte";
</script>\n`
      const output = `<script>
  import Airplane from "phosphor-svelte/lib/Airplane";
import IconGear from "phosphor-svelte/lib/Gear";
</script>\n`

      const result = processor.script({ content, filename: "App.svelte" })

      expect(result.code).toBe(output)
    })
  })
})
