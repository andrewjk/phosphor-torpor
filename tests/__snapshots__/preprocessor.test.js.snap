// Vitest Snapshot v1

exports[`preprocessor > phosphorSvelteOptimize > should preprocess icon imports 1`] = `
{
  "code": "<script>
  import Airplane from \\"phosphor-svelte/lib/Airplane\\";
  import Cube from \\"phosphor-svelte/lib/Cube\\";
import Rectangle from \\"phosphor-svelte/lib/Rectangle\\";
import Horse from \\"phosphor-svelte/lib/Horse\\";
</script>
",
  "dependencies": [],
  "map": SourceMap {
    "mappings": "AAAA,CAAC,MAAM,CCAP;AACA;AACA,EAAE;AAAA;AAAA,8CAAyD;ADC3D,CAAC,CAAC,MAAM;",
    "names": [],
    "sources": [
      "App.svelte",
      "",
    ],
    "version": 3,
  },
  "toString": [Function],
}
`;

exports[`preprocessor > phosphorSvelteOptimize > should preprocess multiple imports 1`] = `
{
  "code": "<script>
  import Airplane from \\"phosphor-svelte/lib/Airplane\\";
import Gear from \\"phosphor-svelte/lib/Gear\\";
  import Cube from \\"phosphor-svelte/lib/Cube\\";
import Rectangle from \\"phosphor-svelte/lib/Rectangle\\";
import Horse from \\"phosphor-svelte/lib/Horse\\";
  import X from \\"phosphor-svelte/lib/X\\";
</script>
",
  "dependencies": [],
  "map": SourceMap {
    "mappings": "AAAA,CAAC,MAAM,CCAP;AACA,EAAE;AAAA,4CAAiD;AACnD,EAAE;AAAA;AAAA,8CAAyD;AAC3D,EAAE,sCAAoC;ADCtC,CAAC,CAAC,MAAM;",
    "names": [],
    "sources": [
      "App.svelte",
      "",
    ],
    "version": 3,
  },
  "toString": [Function],
}
`;

exports[`preprocessor > phosphorSvelteOptimize > should separate type imports 1`] = `
{
  "code": "<script lang=\\"ts\\">
  import Airplane from \\"phosphor-svelte/lib/Airplane\\";
import Gear from \\"phosphor-svelte/lib/Gear\\";
import type { IconContextProps } from \\"phosphor-svelte\\";
</script>
",
  "dependencies": [],
  "map": SourceMap {
    "mappings": "AAAA,CAAC,MAAM,CAAC,IAAI,CAAC,CAAC,EAAE,CAAC,CCAjB;AACA,EAAE;AAAA;AAAA,wDAAwE;ADC1E,CAAC,CAAC,MAAM;",
    "names": [],
    "sources": [
      "App.svelte",
      "",
    ],
    "version": 3,
  },
  "toString": [Function],
}
`;
